docker/resource_docker_service_funcs.go

```diff
@@ -0,0 +1,883 @@
+		}
+	}
+}
+
+func getActiveNodes(ctx context.Context, client *dc.Client) (map[string]struct{}, error) {
+	// nodes, err := client.NodeList(ctx, types.NodeListOptions{})
+	nodes, err := client.ListNodes(dc.ListNodesOptions{Context: ctx})
+	if err != nil {
+		return nil, err
+	}
+
+	activeNodes := make(map[string]struct{})
+	for _, n := range nodes {
+		if n.Status.State != swarm.NodeStateDown {
+			activeNodes[n.ID] = struct{}{}
+		}
+	}
+	return activeNodes, nil
+}
+
+type progressUpdater interface {
+	update(service *swarm.Service, tasks []swarm.Task, activeNodes map[string]struct{}, rollback bool) (bool, error)
+}
+
+type replicatedProgressUpdater struct {
+	// used for mapping slots to a contiguous space
+	// this also causes progress bars to appear in order
+	slotMap map[int]int
+
+	initialized bool
+	done        bool
+}
+
+func (u *replicatedProgressUpdater) update(service *swarm.Service, tasks []swarm.Task, activeNodes map[string]struct{}, rollback bool) (bool, error) {
+	if service.Spec.Mode.Replicated == nil || service.Spec.Mode.Replicated.Replicas == nil {
+		return false, fmt.Errorf("no replica count")
+	}
+	replicas := *service.Spec.Mode.Replicated.Replicas
+
+	if !u.initialized {
+		u.slotMap = make(map[int]int)
+		u.initialized = true
+	}
+
+	tasksBySlot := u.tasksBySlot(tasks, activeNodes)
+
+	// If we had reached a converged state, check if we are still converged.
+	if u.done {
+		for _, task := range tasksBySlot {
+			if task.Status.State != swarm.TaskStateRunning {
+				u.done = false
+				break
+			}
+		}
+	}
+
+	running := uint64(0)
+
+	for _, task := range tasksBySlot {
+		mappedSlot := u.slotMap[task.Slot]
+		if mappedSlot == 0 {
+			mappedSlot = len(u.slotMap) + 1
+			u.slotMap[task.Slot] = mappedSlot
+		}
+
+		if !terminalState(task.DesiredState) && task.Status.State == swarm.TaskStateRunning {
+			running++
+		}
+	}
+
+	if !u.done {
+		log.Printf("[INFO] ... progress: [%v/%v] - rollback: %v", running, replicas, rollback)
+		if running == replicas {
+			log.Printf("[INFO] DONE: all %v replicas running", running)
+			u.done = true
+		}
+	}
+
+	return running == replicas, nil
+}
+
+func (u *replicatedProgressUpdater) tasksBySlot(tasks []swarm.Task, activeNodes map[string]struct{}) map[int]swarm.Task {
+	// If there are multiple tasks with the same slot number, favor the one
+	// with the *lowest* desired state. This can happen in restart
+	// scenarios.
+	tasksBySlot := make(map[int]swarm.Task)
+	for _, task := range tasks {
+		if numberedStates[task.DesiredState] == 0 || numberedStates[task.Status.State] == 0 {
+			continue
+		}
+		if existingTask, ok := tasksBySlot[task.Slot]; ok {
+			if numberedStates[existingTask.DesiredState] < numberedStates[task.DesiredState] {
+				continue
+			}
+			// If the desired states match, observed state breaks
+			// ties. This can happen with the "start first" service
+			// update mode.
+			if numberedStates[existingTask.DesiredState] == numberedStates[task.DesiredState] &&
```

Comment:
I feel like the `numberedStates` map adds indirection that makes it harder to follow what's being checked here. `numberedStates` itself needs to be documented at it's declaration at top explaining that they are proceeded in the order given.

Reply:
ok I'll clarify that

Reply:
I added comment and clarification why I need those `numberedStates`